### DISCUSSION TITLE
Is it possible to use it with Server-Sent Events?

### DISCUSSION THREAD

**Discussion by sandros94** [posva/pinia-colada#118] [Questions] [answered]
In my long journey into learning Pinia I've got to admit that Pinia-Colada did make things much easier, but I'm wondering about a use-case that involves [Server-Sent Events](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events) (in particular in a Nuxt application using [Nitro's SSEs](https://nitro.build/guide/websocket#server-sent-events-sse)), and I'm wondering if I'm going completely offtrack...

My main curiosity point on this is that I always want the freshest state for a particular client's ref, while also using the already available API routes to mutate those states. This means that server-to-client is going to reuse the existing connection, while each client-to-server operation will be a discrete connection (as such, WebSockets will be for a different discussion).

My first idea would be to use [`useEventSource` from Vueuse](https://vueuse.org/core/useEventSource/), but `useQuery`'s `query` param requires an async function, just like an initialization step 🤔

**Comment by posva** [comment]
If you have a realtime connection, you can likely go without Pinia Colada. That being said, it should be possible to use a realtime connection and benefit from optimistic updates. I haven't tried it myself but if I had to implement it, I would create a custom `useRealtimeQuery()` (using `useEventSource()` internally) that wraps `useQuery()` and instantiates the connection. The `query` function would then just await for the initial data and then resolve immediately. If you come up with anything, I'm curious about it. It might be useful to add it as a cookbook entry or even just another composable

**Reply by sandros94** [reply]
Assuming I'm building a composable, as I was reading the [_Reusable Queries_](https://pinia-colada.esm.dev/guide/queries.html#Reusable-Queries), I should work with `defineQuery()` instead of `useQuery()` right? Since `useEventSource()` does return a ref and could make it out of sync if reused in different pages/components?

Sorry if this question may sound dumb, I'm mainly testing the ground

**Reply by sandros94** [reply]
So far I've tested this:
```ts
import { joinURL } from 'ufo'

export const useRealtimeQuery = defineQuery(() => {
  const { origin } = useRequestURL()
  const queryCache = useQueryCache()
  const { status, event, data } = useEventSource(joinURL(origin, '/api/test/event'), ['test'])

  const { state, ...rest } = useQuery({
    key: ['events', { event: event.value }],
    query: async () => {
      await new Promise<void>((resolve) => {
        if (data.value) {
          resolve()
        }
        else {
          const unwatch = watch(data, () => {
            unwatch()
            resolve()
          })
        }
      })

      return data.value
    },
  })

  watch(data, async () => {
    if (status.value === 'OPEN' && data.value) {
      console.log('New data:', data.value)
      await queryCache.invalidateQueries({ key: ['events', { event: event.value }] })
    }
  })

  return {
    ...rest,
    state,
  }
})
```

Initialization work as intended, but I'm not sure if the `invalidateQueries` should be used like this, since the state isn't being updated

**Reply by posva** [reply]
Since you are writing a composable, you don't need `defineQuery()`, you need `useQuery()` within your own composable:

```ts
export function useRealtimeQuery(opts) {
  const query = useQuery(...)
  // ...
}
```

I wouldn't expose the returned ref from `useEventSource()`, and use the query cache as the source of truth instead

**Reply by posva** [reply]
The invalidateQueries doesn't help much since you have realtime data. You should use something like `setQueryData()` to update the cache.
BTW, you shouldn't call `watch` within the query, that can potentially create a memory leak (same as calling it within a method)

**Reply by sandros94** [reply]
My current working approach:
```ts
import { joinURL } from 'ufo'
import { safeDestr } from 'destr'

export function useRealtimeQuery(events?: string | string[]) {
  const _events = events
    ? Array.isArray(events)
      ? events
      : [events]
    : undefined

  const { event, data } = useEventSource(joinURL(useRequestURL().origin, '/api/test/event'), _events)
  const { state, ...rest } = useQuery({
    key: () => ['SSE', { event: event.value }],
    query: () => new Promise<string | null>((resolve) => {
      resolve(safeDestr(data.value))
    }),
  })

  watch(data, (newData) => {
    useQueryCache().setQueryData(
      ['SSE', { event: event.value }],
      safeDestr(newData),
    )
  })

  return {
    ...rest,
    state,
  }
}
```

Am I doing things alright? I could see your point on going without Pinia Colada with this use-case

**Reply by posva** [reply, marked as answer]
I think it's something more like

```ts
import { joinURL } from 'ufo'
import { safeDestr } from 'destr'

export function useRealtimeQuery(events?: string | string[]) {
  const _events = events
    ? Array.isArray(events)
      ? events
      : [events]
    : undefined

  let resolve: null | (() => void), reject: null | ((error: any) => void)

  const queryPromise = new Promise<void>((res, rej) => {
    resolve = res
    reject = rej
  })

  const { event, data } = useEventSource(joinURL(useRequestURL().origin, '/api/test/event'), _events)
  const queryCache = useQueryCache()

  watch(data, (newData) => {
    // initial promise
    if (resolve) {
      resolve()
      resolve = null
    } else {
      // subsequently we only need to update the cache
      queryCache.setQueryData(
        ['SSE', { event: event.value }],
        safeDestr(newData),
      )
    }
  })

  return useQuery({
    key: () => ['SSE', { event: event.value }],
    async query() {
      await queryPromise
      return safeDestr(data.value)
    }),
  })
}
```

- Composables outside of `watch`
- Resolve only once
- There is no error handling (reject is not used)
- The `key` should probably be the `events` and be required to ensure unicity

**Reply by sandros94** [reply]
> ```ts
> let resolve: null | (() => void), reject: null | ((error: any) => void)
>
> const queryPromise = new Promise<void>((res, rej) => {
>   resolve = res
>   reject = rej
> })
> ```

This felt like you were actually writing stuff and I'm still playing with Duplo 😅

> - The `key` should probably be the `events` and be required to ensure unicity

Indeed, currently any global events will override each other

> - There is no error handling (reject is not used)

It might be just as simple as watching `error` ref returned by `useEventSource` and passing it to the reject?

Thank you so much fo guiding me 🙏

**Reply by posva** [reply]
You are being too hard on yourself 🤣 
Watching the error seems good

**Comment by dimroc** [comment]
Hi all,
Thanks for asking this @sandros94 , this is still relevant a year later. It would be great to have recommendations around server-pushed data. [TanStack has some documentation](https://github.com/TanStack/query/blob/3bcaa417ac302a93d17c4287d518121bb77d1b26/docs/framework/react/community/tkdodos-blog.md#7-using-websockets-with-react-query) around query invalidation and [partial updates](https://tkdodo.eu/blog/using-web-sockets-with-react-query#partial-data-updates). It could be as simple as "don't use them together."
